### PR TITLE
feat(nexus): nexus create snapshot api

### PR DESCRIPTION
**Pull request by mtzaurus**
Added CreateSnapshot method to V1 Nexus gRPC API.

### CODE REVIEW

**Review by hrudaya21 — APPROVED** [review]

File: protobuf/v1/nexus.proto [hunk 1]
```diff
@@ -67,6 +70,18 @@ message CreateNexusRequest {
...
+// Create nexus snapshot.
+message  CreateSnapshotRequest {
+  string nexus_uuid = 1;    // UUID of the nexus
+  string snapshot_name = 2; // User-specific snapshot name.
+  string txn_id = 3;        // unique transaction id generate for snapshot creation
+}
+
+// Nexus snapshot creation response.
+message CreateSnapshotResponse {
+  string snapshot_uuid = 1; // UUID of the nexus snapshot
```
Comment (tiagolobocastro) [inline protobuf/v1/nexus.proto:82]: what's the difference between txn_id and snapshot_uuid here?

Comment (mtzaurus) [reply protobuf/v1/nexus.proto:82]: Snapshot UUID is the unique UUID of the snapshot, like UUID of the device.
TX_ID is not supposed the be unique, it's part of the key (entity_id, tx_id), an this combination is supposed to be unique.
So we return UUID (which is generated by SPDK).

Comment (tiagolobocastro) [reply protobuf/v1/nexus.proto:82]: do we need to also specify this uuid as an argument as well then so we can make the call idempotent? I think for the lvol for example we use methods with allow us to specify a UUID selected by the control-plane.

Comment (mtzaurus) [reply protobuf/v1/nexus.proto:82]: OK, we could have that if that's needed to the Control Plane.

Comment (hrudaya21) [reply protobuf/v1/nexus.proto:82]: For Replica Snapshot We are following the same as nexus.. We get the txn-id from control plane and return UUID from dataplane which is generated by SPDK.

// Create Replica Snapshot Request.
message CreateReplicaSnapshotRequest {
  string replica_uuid = 1;  // uuid of the replica
  string snapshot_name = 2; // snapshot to be created
  string txn_id = 3; // unique transaction id generate for snapshot creation
}

/// Create Replica Snapshot Response.
message CreateReplicaSnapshotResponse {
  string snapshot_uuid = 1; // UUID of the snapshot newly created
}

File: protobuf/v1/nexus.proto [hunk 2]
```diff
@@ -67,6 +70,18 @@ message CreateNexusRequest {
   NexusNvmePreemption preemptPolicy = 11; // the preemption policy to use
 }
 
+// Create nexus snapshot.
+message  CreateSnapshotRequest {
+  string nexus_uuid = 1;    // UUID of the nexus
+  string snapshot_name = 2; // User-specific snapshot name.
```
Comment (dsharma-dc) [inline protobuf/v1/nexus.proto:76]: Should the name be optional? If a user isn't coming up with a name, we internally create a name as a string concat of other params and timestamp for example.

**Review by dsharma-dc — APPROVED** [review]

File: protobuf/v1/nexus.proto (hunk 2 above)
Comment (hrudaya21) [reply protobuf/v1/nexus.proto:76]: > Should the name be optional? If a user isn't coming up with a name, we internally create a name as a string concat of other params and timestamp for example.

I think name should be generated by control plane if the information is not coming as part of the CSI gRPC. For dataplane, we can consider the parameter with valid value.

Comment (dsharma-dc) [reply protobuf/v1/nexus.proto:76]: Alright, yes since this is gRPC spec between control-plane and data-plane, control-plane should take are of that.

**Comment by mtzaurus** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
@mtzaurus there was an invalid setting on the protection. You should be able to use the UI to merge now.

**Review by tiagolobocastro — APPROVED** [review]

**Pull request merged**